src/filter/index.js

```diff
@@ -17,10 +17,79 @@ export default async function filter(content, selector, filterNames, filterFunct
     });
   }
 
-  const selectedContent = Array.from(webPageDOM.querySelectorAll(selector));
-  if (!selectedContent.length) {
-    console.warn(`The provided selector "${selector}" has no match in the web page.`);
+  let selectedContent = webPageDOM;
+
+  if (removeElements) {
+    let elementsToRemove = removeElements;
+
+    if (!Array.isArray(elementsToRemove)) {
+      elementsToRemove = [elementsToRemove];
+    }
+
+    elementsToRemove.forEach((elementSelector) => {
+      if (typeof elementSelector === 'object') {
+        const rangeSelection = getRangeSelection(selectedContent, elementSelector);
+        rangeSelection.deleteContents();
+      } else {
+        Array.from(webPageDOM.querySelectorAll(elementSelector)).forEach(node => node.remove());
+      }
+    });
+  }
+
+  const selectedContents = [];
+  if (selector) {
+    let elementsToSelect = selector;
+
+    if (!Array.isArray(elementsToSelect)) {
+      elementsToSelect = [elementsToSelect];
+    }
+
+    elementsToSelect.forEach((elementSelector) => {
+      if (typeof elementSelector === 'object') {
+        const rangeSelection = getRangeSelection(selectedContent, elementSelector);
+        selectedContents.push(rangeSelection.cloneContents());
+      } else {
+        selectedContents.push(...Array.from(webPageDOM.querySelectorAll(elementSelector)));
+      }
+    });
+  }
+
+  if (!selectedContents.length) {
+    throw new Error(`The provided selector "${selector}" has no match in the web page.`);
+  }
+
+  return selectedContents.map(domFragment => turndownService.turndown(domFragment)).join('\n');
+}
+
+function getRangeSelection(document, rangeSelector) {
+  const { startBefore, startAfter, endBefore, endAfter } = rangeSelector;
+
+  if (startBefore && startAfter) {
+    throw new Error(`Selectors "startBefore" and "startAfter" in range selector cannot both be defined. Specify only one in: ${JSON.stringify(rangeSelector)}`);
+  }
+
+  if (endBefore && endAfter) {
+    throw new Error(`Selectors "endBefore" and "endAfter" in range selector cannot both be defined. Specify only one in: ${JSON.stringify(rangeSelector)}`);
+  }
+
+  if (!((startBefore || startAfter) && (endBefore || endAfter))) {
+    throw new Error(`At least one "start" and one "end" should be defined in: ${JSON.stringify(rangeSelector)}`);
```

Comment:
👏 now that's excellent error checking!

However, I believe this would be better checked statically in the JSON Schema 🙂